index b3f9c6a4..3fb63b82
--- a/perl/bsvf/cmp/vfs-2016-08-17.r2/include/ReadPreprocess.pm
+++ b/perl/bsvf/cmp/vfs-2016-08-17.r2/include/ReadPreprocess.pm
@@ -202,7 +202,7 @@ sub dynamic_trim_fastq_tag { #(File_in, phred_scale, desiredQ, emit_threshold)
 	my $is_gzip_input = DetermineFileType ($file_in);
 
 	if ($is_gzip_input) {
-		open ($FH,"-|", "gzip -dc $file_in") or die "Can't open $file_in (gzipped) for reading : $!";
+		open ($FH, "<:gzip", $file_in) or die "Can't open $file_in (gzipped) for reading : $!";
 	}
 	else {
 		open $FH, '<', $file_in or die "Can't open $file_in for reading : $!";

index 73005d03..30013ace 100644
--- a/perl/bsvf/bin/BSseeker2/bs_utils/utils.py
+++ b/perl/bsvf/bin/BSseeker2/bs_utils/utils.py
@@ -272,7 +272,7 @@ def read_fasta(fasta_file):
         print "[Error] Cannot find fasta file : %s !" % fasta_file
         exit(-1)
     sanitize = re.compile(r'[^ACTGN]')
-    sanitize_seq_id = re.compile(r'[^A-Za-z0-9]')
+    sanitize_seq_id = re.compile(r'[^A-Za-z0-9|]')
 
     chrom_seq = []
     chrom_id = None

